File: python/examples/harness/scenarios/vector.py
"""Scenarios for vector search degradation and sqlite-vec insert/search."""

from __future__ import annotations

from fathomdb import (
    ChunkInsert,
    ChunkPolicy,
    NodeInsert,
    ProjectionTarget,
    VecInsert,
    WriteRequest,
    new_row_id,
)

from ..models import (
    HarnessContext,
    ScenarioResult,
    VECTOR_CHUNK_ID,
    VECTOR_DOCUMENT_ID,
    VECTOR_QUERY,
    VECTOR_SOURCE,
)


def vector_degradation(context: HarnessContext) -> ScenarioResult:
    """Validate that vector search gracefully degrades in baseline mode."""
    # Stay on the legacy `vector_search` step (the new `vector_search()`
    # Python shim is a deprecation route into semantic_search, which goes
    # through a compile path that does not yet return was_degraded on a
    # missing vec table — see Pack F2 notes).
    rows = (
        context.engine.nodes("Document")
        ._with_step({"type": "vector_search", "query": VECTOR_QUERY, "limit": 3})
        .execute()
    )
    assert rows.was_degraded is True, "baseline vector query should degrade"
    assert rows.nodes == [], (
        f"expected no vector rows, got {[node.logical_id for node in rows.nodes]}"
    )
    return ScenarioResult(name="vector_degradation")


def vector_insert_and_search(context: HarnessContext) -> ScenarioResult:
    """Validate vector insert and nearest-neighbor search returns the expected node."""
    context.engine.write(
        WriteRequest(
            label="vector-insert-and-search",
            nodes=[
                NodeInsert(
                    row_id=new_row_id(),
                    logical_id=VECTOR_DOCUMENT_ID,
                    kind="Document",
                    properties={"title": "Vector document"},
                    source_ref=VECTOR_SOURCE,
                    upsert=True,
                    chunk_policy=ChunkPolicy.REPLACE,
                )
            ],
            chunks=[
                ChunkInsert(
                    id=VECTOR_CHUNK_ID,
                    node_logical_id=VECTOR_DOCUMENT_ID,
                    text_content="vector-enabled retrieval payload",
                )
            ],
            vec_inserts=[
                VecInsert(
                    chunk_id=VECTOR_CHUNK_ID,
                    embedding=[0.1, 0.2, 0.3, 0.4],
                )
            ],
        )
    )

    rows = (
        context.engine.nodes("Document")
        ._with_step({"type": "vector_search", "query": VECTOR_QUERY, "limit": 5})
        .execute()
    )
    assert rows.was_degraded is False, "vector query should not degrade in vector mode"
    assert any(node.logical_id == VECTOR_DOCUMENT_ID for node in rows.nodes), (
        f"expected {VECTOR_DOCUMENT_ID} in vector results, got {[node.logical_id for node in rows.nodes]}"
    )

    repair = context.engine.admin.rebuild(target=ProjectionTarget.VEC)
    assert repair.targets == [ProjectionTarget.VEC], (
        f"unexpected targets={repair.targets}"
    )
    semantics = context.engine.admin.check_semantics()
    assert semantics.stale_vec_rows == 0, f"stale_vec_rows={semantics.stale_vec_rows}"
    assert semantics.vec_rows_for_superseded_nodes == 0, (
        f"vec_rows_for_superseded_nodes={semantics.vec_rows_for_superseded_nodes}"
    )
    return ScenarioResult(name="vector_insert_and_search")

File: typescript/apps/sdk-harness/src/scenarios/vector.ts
import { Engine, WriteRequestBuilder, newId, newRowId } from "fathomdb";
import type { HarnessResult } from "../models.js";
import { assert, handleScenarioError, tempDbPath } from "../skip.js";

const VECTOR_DIM = 4;

export function vectorScenario(): HarnessResult {
  try {
    const engine = Engine.open(tempDbPath("vector"), { vectorDimension: VECTOR_DIM });

    const builder = new WriteRequestBuilder("vector-ingest");
    const node = builder.addNode({
      rowId: newRowId(), logicalId: newId(), kind: "Embedding",
      properties: { topic: "machine learning" }, sourceRef: "vector-test",
    });
    const chunk = builder.addChunk({
      id: newId(), node, textContent: "Neural networks for classification.",
    });
    builder.addVecInsert({
      chunk,
      embedding: [0.1, 0.2, 0.3, 0.4],
    });
    const receipt = engine.write(builder.build());
    assert(receipt.warnings.length === 0, "unexpected write warnings");

    const rows = engine.nodes("Embedding")
      .vectorSearch("classification", 5)
      .execute();
    assert(rows.nodes.length >= 1, `expected ≥1 vector result, got ${rows.nodes.length}`);
    assert(rows.nodes[0].kind === "Embedding", "kind mismatch");

    const standard = engine.nodes("Embedding").execute();
    assert(standard.nodes.length >= 1, "standard query should return results");

    engine.close();
    return { name: "vector", ok: true };
  } catch (error) {
    return handleScenarioError("vector", error);
  }
}
